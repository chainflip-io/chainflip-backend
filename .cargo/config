# Note: If you just want to save typing command commands, you can install tab completions for most shells. Type
# `rustup completions --help` to find out how to set this up.

[alias]
# run unit tests but not integration tests
cf-test = "test --lib --all-features"
# TODO: add all-features and clean up benchmarks.
cf-clippy = "clippy --all-targets"

ci-check = "check --all-targets --all-features --release"
ci-build = "build --release --features runtime-benchmarks"
ci-test = "test --lib --all-features --release"
ci-clippy = "clippy --all-targets -- -D warnings"
