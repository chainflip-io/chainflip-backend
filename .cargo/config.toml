[env]
CF_ETH_CONTRACT_ABI_ROOT = { value = "eth-contract-abis", relative = true }
CF_ETH_CONTRACT_ABI_TAG = "v1.0.0"
CF_TEST_CONFIG_ROOT = { value = "engine/config/testing", relative = true }

# Note: If you just want to save typing command commands, you can install tab completions for most shells. Type
# `rustup completions --help` to find out how to set this up.

[alias]
cf-test = "test --lib --features runtime-benchmarks,try-runtime"
cf-test-ci = "test --lib --features runtime-benchmarks,slow-tests,try-runtime"
cf-test-cfe = "test --lib --package chainflip-engine --package multisig"

cf-clippy = "clippy --all-targets --features runtime-benchmarks,try-runtime,runtime-integration-tests,slow-tests -- -D warnings"
cf-clippy-ci = "clippy --all-targets --features runtime-benchmarks,try-runtime,runtime-integration-tests,slow-tests -- -D warnings"

cf-build = "build --features runtime-benchmarks,try-runtime"

cf-build-try-runtime = "build --release --features try-runtime"
cf-build-release = "build --release"
cf-build-production = "build --profile=production"
# Check for feature inconsistencies.
check-sc-features = '''
tree --no-default-features --depth 1 --edges=features,normal
    -p state-chain-*
    -p pallet-cf-*
    -p cf-*
'''

# The `cf-audit` command is configured to ignore the following advisories:
# Vulnerabilities:
# - RUSTSEC-2022-0093: This advisory is related to ed25519-dalek's public api design, which we don't expose.
# - RUSTSEC-2023-0052: This advisory comes from webpki, a dependency of ethers-rs. CPU denial of service in certificate path building. We don't use this functionality.
# Unsound:
# - RUSTSEC-2021-0145: This advisory only affects the windows platform. It's a transitive dependency of `substrate`.
# Unmaintained:
# - RUSTSEC-2020-0168: This advisory comes from `mach`, which is unmaintained but not a security issue. It's a dependency of `subxt`.
# - RUSTSEC-2021-0059: This is a transitive dependency of libp2p and will be fixed in an upcoming release.
# - RUSTSEC-2021-0060: This is a transitive dependency of libp2p and will be fixed in an upcoming release.
# - RUSTSEC-2021-0139: This advisory comes from `ansi_term`, which is unmaintained but not a security issue. It's a dependency of `subxt`.
# - RUSTSEC-2022-0061: This advisory is related to the deprecated `parity-wasm`, not a security issue. It's a dependency of `substrate`.
# - RUSTSEC-2023-0074: Related to zerocopy, this is a transitive dependency. We can update this when we update substrate.
cf-audit = '''
audit -D unmaintained -D unsound
    --ignore RUSTSEC-2022-0093
    --ignore RUSTSEC-2023-0052
    --ignore RUSTSEC-2021-0145
    --ignore RUSTSEC-2020-0168
    --ignore RUSTSEC-2021-0059
    --ignore RUSTSEC-2021-0060
    --ignore RUSTSEC-2021-0139
    --ignore RUSTSEC-2022-0061
    --ignore RUSTSEC-2023-0074
'''
