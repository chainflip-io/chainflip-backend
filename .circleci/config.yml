######################## Anchors ########################
only_main_branches: &only_main_branches
  filters:
    branches:
      only:
        - main
        - develop
        - /epic\/.*/
        - /release\/.*/
        # - update/epic-ci-1
ignore_main_branches: &ignore_main_branches
  filters:
    branches:
      ignore:
        - main
        - develop
        - /epic\/.*/
        - /release\/.*/
        # - update/epic-ci-1

auth: &dockerconfig
  username: $DOCKER_USERNAME
  password: $DOCKER_PASSWORD

cargo-base-job: &cargo-base-job
  executor: rust-base
  environment:
    SCCACHE_CACHE_SIZE: 14G
  resource_class: xlarge

######################## /Anchors ########################

######################## Config ########################
version: 2.1
executors:
  chainflip-master:
    docker:
      - image: ghcr.io/chainflip-io/chainflip-platform-monorepo/chainflip-master:latest
        auth: *dockerconfig
  rust-base:
    docker:
      - image: ghcr.io/chainflip-io/chainflip-backend/rust-base:nightly-2022-08-08
        auth: *dockerconfig
  rust-base-ci:
    docker:
      - image: ghcr.io/chainflip-io/chainflip-backend/rust-base:nightly-2022-08-08
        auth: *dockerconfig
        user: ci
  java-executor:
    docker:
      - image: cimg/openjdk:17.0.2
        auth: *dockerconfig
orbs:
  docker: circleci/docker@1.7.0
  discord: antonioned/discord@0.1.0
  gh: circleci/github-cli@1.0.4
  doctl: digitalocean/cli@0.1.1
  helm: circleci/helm@1.2.0
parameters:
  rust-base:
    type: boolean
    default: false
  build_testnet:
    type: boolean
    default: false
######################## /Config ########################

######################## Commands ########################
commands:
  install_gh_cli:
    parameters:
      version:
        default: 2.4.0
        type: string
    steps:
      - run: |
          wget https://github.com/cli/cli/releases/download/v<<parameters.version>>/gh_<<parameters.version>>_linux_amd64.tar.gz
          tar -xvf gh_<<parameters.version>>_linux_amd64.tar.gz
          mv ./gh_<<parameters.version>>_linux_amd64/bin/gh /usr/local/bin
  send_notification:
    parameters:
      failure_message:
        type: string
        default: ":red_circle: Pipeline failed for branch `<<pipeline.git.revision>>`"
      fail_only:
        type: boolean
        default: true
      only_for_branches:
        type: string
        default: "develop,main"
    steps:
      - discord/status:
          fail_only: <<parameters.fail_only>>
          failure_message: <<parameters.failure_message>>
          only_for_branches: <<parameters.only_for_branches>>
          webhook: $DW_ALERTS_STAGING
  merge_target:
    steps:
      - run:
          name: Merge target into your branch
          command: |
            # setup the github user
            git config --global user.email $( git log --format='%ae' $CIRCLE_SHA1^! )
            git config --global user.name $( git log --format='%an' $CIRCLE_SHA1^! )

            # if this is a PR, we extract the API URL to grab the PR as JSON from the Github API
            if [[ -n "$CIRCLE_PULL_REQUEST" ]]; then
              url="$(echo "$CIRCLE_PULL_REQUEST" | sed 's/github.com/api.github.com\/repos/' | sed 's/pull/pulls/')"
              target_branch=$(
                curl -H "Authorization: token $GITHUB_TOKEN" "$url" | jq '.base.ref' | tr -d '"'
              )
              echo "$target_branch"

              # this will fail if $target_branch is null
              # if it should pass, wrap this in `if [[ -n "$target_branch" ]]; then`
              git fetch && git merge origin/$target_branch --no-edit
            fi
######################## /Commands ########################

######################## Workflows ########################
workflows:
  chainflip-backend-non-blocking:
    jobs:
      - cargo-doc:
          context:
            - ghcr-credentials
            - github-credentials
      - update-cargo-audit-log:
          context:
            - ghcr-credentials
            - github-credentials
          <<: *only_main_branches

  chainflip-backend-checks:
    jobs:
      - cargo-clippy:
          requires:
            - cargo-fmt
          context:
            - ghcr-credentials
            - github-credentials
          <<: *ignore_main_branches

      # The follow jobs always run, regardless of PR or merge.
      - cargo-check:
          requires:
            - cargo-fmt
          context:
            - ghcr-credentials
            - github-credentials
      - cargo-fmt:
          context:
            - ghcr-credentials
            - github-credentials
      - cargo-build-and-test:
          requires:
            - cargo-fmt
          context:
            - ghcr-credentials
            - github-credentials
            - discord-webhooks
      - git-clone-docgen-repo:
          requires:
            - cargo-build-and-test
          context:
            - github-credentials
      - test-single-node:
          requires:
            - git-clone-docgen-repo
          context:
            - ghcr-credentials
            - infura-endpoints
      - docgen-store-artifacts:
          requires:
            - test-single-node
          context:
            - do-credentials

######################## /Workflows ########################

######################## Jobs ########################
jobs:
  docker-build:
    executor: docker/docker
    parameters:
      image:
        type: string
        description: Name of primary tag without the registry
        default: ""
      extra_build_args:
        type: string
        description: Any extra arguments to pass to the build
        default: ""
      dockerfile:
        type: string
        description: Location of the Dockerfile
        default: Dockerfile
      registry:
        type: string
        description: Registry of Docker images
        default: ghcr.io
      tag:
        type: string
        description: Primary tag of the build
        default: latest,<<pipeline.git.revision>>
      cache_from:
        type: string
        description: Where to derive the build cache
        default: ""
    steps:
      - setup_remote_docker:
          version: 20.10.7
          docker_layer_caching: true
      - checkout
      - attach_workspace:
          at: ./
      - docker/check:
          docker-username: DOCKER_USERNAME
          registry: ghcr.io
      - docker/build:
          image: <<parameters.image>>
          registry: ghcr.io
          extra_build_args: <<parameters.extra_build_args>>
          dockerfile: <<parameters.dockerfile>>
          cache_from: <<parameters.cache_from>>
          tag: <<parameters.tag>>
      - docker/push:
          registry: ghcr.io
          image: <<parameters.image>>
          tag: <<parameters.tag>>

  cargo-build-and-test:
    executor: rust-base-ci
    resource_class: xlarge
    steps:
      - checkout
      - merge_target
      - run: sccache --show-stats
      - run: |
          cargo build --locked --release --features runtime-benchmarks
          sccache --show-stats
      - run: |
          cargo test --lib --locked --release --all-features
          sccache --show-stats
      - persist_to_workspace:
          root: ./
          paths:
            - target/release/chainflip-node
            - target/release/chainflip-engine
            - target/release/chainflip-cli
            - target/release/generate-genesis-keys
            - target/release/wbuild/state-chain-runtime/state_chain_runtime.wasm
      - send_notification:
          only_for_branches: "develop"

  cargo-fmt:
    executor: rust-base
    steps:
      - checkout
      - merge_target
      - run: >
          cargo fmt --all -- --check

  # Check and test the entire chainflip-backend.
  # We run sccache --show-stats multiple times;
  # First to check connection, and once after each
  # cargo function to track the performance of the compilation.
  cargo-check:
    <<: *cargo-base-job
    steps:
      - checkout
      - merge_target
      - run: sccache --show-stats
      - run:
          environment:
            RUSTFLAGS: -D warnings
          command: |
            cargo check --all-targets --all-features --locked --release
            sccache --show-stats

  cargo-test:
    <<: *cargo-base-job
    steps:
      - checkout
      - merge_target
      - run: sccache --show-stats
      - run: |
          cargo test --lib --locked --release
          sccache --show-stats

  cargo-clippy:
    <<: *cargo-base-job
    steps:
      - checkout
      - merge_target
      - run: sccache --show-stats
      # TODO: remove the global ignores (#1254)
      - run: |
          sh clippy.sh
          sccache --show-stats

  cargo-doc:
    <<: *cargo-base-job
    shell: /bin/bash # Circle's beautiful way of ignoring failures ðŸ™„
    steps:
      - checkout
      - merge_target
      - run: sccache --show-stats
      - run: |
          cargo doc --workspace --document-private-items --no-deps
          sccache --show-stats
      - store_artifacts:
          path: target/doc/

  update-cargo-audit-log:
    <<: *cargo-base-job
    shell: /bin/bash # Circle's beautiful way of ignoring failures ðŸ™„
    steps:
      - checkout
      - merge_target
      - install_gh_cli
      - run: sccache --show-stats
      - run: |
          cargo install --force cargo-audit
          cargo generate-lockfile
      - run: |
          apt-get install -y gettext
          wget https://github.com/cli/cli/releases/download/v2.4.0/gh_2.4.0_linux_amd64.tar.gz
          tar -xvf gh_2.4.0_linux_amd64.tar.gz
          cargo-audit audit --json > audit.json
          export advisory=$(cat audit.json | jq '.vulnerabilities.list[].advisory | {id,package,title,description,date,url}')
          export warnings=$(cat audit.json | jq '.warnings | keys[] as $k| .[$k] | .[] | {kind,package}')
          export date=$(date)
          cat .circleci/templates/audit-log.md | envsubst > /tmp/audit-log.md
          gh issue edit 1175 --body-file /tmp/audit-log.md

  git-clone-docgen-repo:
    resource_class: small
    docker:
      - image: "cimg/node:17.8.0"
    steps:
      - gh/setup
      - gh/clone:
          repo: https://github.com/chainflip-io/chainflip-docgen.git
          dir: chainflip-docgen
      - run: |
          cd chainflip-docgen
          npm install
      - persist_to_workspace:
          root: ./
          paths:
            - chainflip-docgen/*

  test-single-node:
    executor: chainflip-master
    steps:
      - checkout
      - attach_workspace:
          at: ./
      - run:
          name: Setup config
          command: |
            mkdir config
            echo -n $BASHFUL_SIGNING_KEY > signing_key_file
            echo -n $ETH_PRIVATE_KEY > eth_private_key_file
            echo -n $NODE_KEY > node_key_file
      - run:
          name: Start Node
          background: true
          command: ./target/release/chainflip-node --chain cf-dev --node-key-file=node_key_file
      - run:
          name: Check Node Health
          command: >
            curl --request POST -L
            -H "Content-Type:application/json;charset=utf-8"
            --data '{"id":1, "jsonrpc":"2.0", "method": "system_health"}'
            --url 'http://localhost:9933'
            --connect-timeout 5
            --max-time 10
            --retry 3
            --retry-delay 1
            --retry-max-time 20
            --retry-all-errors
            -vv
      - run:
          name: Start Engine
          background: true
          command: ./target/release/chainflip-engine --config-path engine/config/CI.toml
      - run:
          name: Check Engine Health
          command: >
            curl
            --url 'http://localhost:5555/health'
            --connect-timeout 5
            --max-time 10
            --retry 3
            --retry-delay 1
            --retry-max-time 20
            --retry-all-errors
            -vv
      - run:
          name: Generate State Chain docs
          command: |
            export WS_ENDPOINT="ws://localhost:9944"
            export TYPES_FILE="../state-chain/types.json"
            cd ./chainflip-docgen/
            npm run generate
      - persist_to_workspace:
          root: ./
          paths:
            - chainflip-docgen/*
      - store_artifacts:
          path: target/release

  docgen-store-artifacts:
    executor: java-executor
    steps:
      - attach_workspace:
          at: ./
      - run:
          name: Install graphviz
          command: |
            sudo apt-get update
            sudo apt-get install -y graphviz
      - run:
          name: Generate class diagram
          command: |
            cd ./chainflip-docgen/
            java -jar plantuml.jar generated/class.puml
      - store_artifacts:
          path: ./chainflip-docgen/generated
######################## /Jobs ########################
