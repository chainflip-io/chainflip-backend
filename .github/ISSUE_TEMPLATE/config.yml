contact_links:
  - name: Infrastructure changes
    url: https://github.com/chainflip-io/chainflip-platform-monorepo/issues/new?assignees=&labels=request%2Ceffort-0&template=003-request.md&title=%5BREQUEST%5D+-+
    about: For any infrastructure changes, please raise an issue in the Platform team's monorepo.
