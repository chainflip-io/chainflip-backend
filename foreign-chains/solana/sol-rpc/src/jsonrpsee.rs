use jsonrpsee::{
	core::{client::ClientT, Error},
	http_client::HttpClient,
};

use crate::traits::{Call, CallApi};

#[async_trait::async_trait]
impl CallApi for HttpClient {
	type Error = Error;
	async fn call<C: Call>(&self, call: C) -> Result<C::Response, Self::Error> {
		self.request(C::CALL_METHOD_NAME, call.call_params())
			.await
			.and_then(|js_value| <C as Call>::process_response(&call, js_value).map_err(Into::into))
	}
}
