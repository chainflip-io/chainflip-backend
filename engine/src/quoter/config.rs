use config::{Config, ConfigError, Environment, File};
use serde::Deserialize;
use std::env;

lazy_static! {
    /// The quoter config
    pub static ref QUOTER_CONFIG: QuoterConfig = match QuoterConfig::new("config/quoter") {
        Ok(config) => config,
        Err(err) => {
            error!("Failed to load config: {}", err);
            panic!("Failed to load config");
        }
    };
}

#[derive(Debug, Deserialize, Clone)]
/// Configuration for database
pub struct Database {
    /// RPC specific config
    pub name: String,
}

#[derive(Debug, Deserialize, Clone)]
/// Configuration for quoter
pub struct QuoterConfig {
    /// Oxen config
    pub database: Database,
    /// The vault node url
    pub vault_node_url: String,
}

impl QuoterConfig {
    fn new(path: &str) -> Result<Self, ConfigError> {
        let mut config = Config::new();

        // Start off by merging in the "default" configuration file
        config.merge(File::with_name(&format!("{}/default", path)))?;

        // Add in the current environment file
        // Default to 'development' env
        // Note that this file is _optional_
        let env = env::var("RUN_MODE").unwrap_or_else(|_| "development".into());
        config.merge(File::with_name(&format!("{}/{}", path, env)).required(false))?;

        // Add in a local configuration file
        // This file shouldn't be checked in to git
        config.merge(File::with_name(&format!("{}/local", path)).required(false))?;

        // Add in settings from the environment (with a prefix of QUOTER)
        // Eg.. `QUOTER_DEBUG=1 ./target/app` would set the `debug` key
        config.merge(Environment::with_prefix("quoter"))?;

        // You can deserialize (and thus freeze) the entire configuration as
        config.try_into()
    }
}
