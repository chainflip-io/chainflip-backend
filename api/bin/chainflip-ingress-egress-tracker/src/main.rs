use cf_chains::{
	btc::BitcoinNetwork, evm::SchnorrVerificationComponents, AnyChain, Bitcoin, Chain, Ethereum,
	Polkadot,
};
use cf_primitives::{Asset, BroadcastId, ForeignChain};
use chainflip_engine::{
	settings::{insert_command_line_option, CfSettings, HttpBasicAuthEndpoint, WsHttpEndpoints},
	state_chain_observer::client::{
		chain_api::ChainApi, storage_api::StorageApi, StateChainClient, STATE_CHAIN_CONNECTION,
	},
};
use clap::Parser;
use config::{Config, ConfigBuilder, ConfigError, Environment, Map, Source, Value};
use futures::FutureExt;
use jsonrpsee::{core::Error, server::ServerBuilder, RpcModule};
use pallet_cf_broadcast::TransactionOutIdFor;
use pallet_cf_ingress_egress::DepositWitness;
use serde::{Deserialize, Serialize};
use state_chain_runtime::PalletInstanceAlias;
use std::{collections::HashMap, env, net::SocketAddr};
use tracing::log;
use utilities::{rpc::NumberOrHex, task_scope};

mod witnessing;

#[derive(Serialize)]
struct WitnessAsset {
	chain: ForeignChain,
	asset: Asset,
}

impl From<cf_chains::assets::eth::Asset> for WitnessAsset {
	fn from(asset: cf_chains::assets::eth::Asset) -> Self {
		match asset {
			cf_chains::assets::eth::Asset::Eth |
			cf_chains::assets::eth::Asset::Flip |
			cf_chains::assets::eth::Asset::Usdc =>
				Self { chain: ForeignChain::Ethereum, asset: asset.into() },
		}
	}
}

impl From<cf_chains::assets::dot::Asset> for WitnessAsset {
	fn from(asset: cf_chains::assets::dot::Asset) -> Self {
		match asset {
			cf_chains::assets::dot::Asset::Dot =>
				Self { chain: ForeignChain::Polkadot, asset: asset.into() },
		}
	}
}

impl From<cf_chains::assets::btc::Asset> for WitnessAsset {
	fn from(asset: cf_chains::assets::btc::Asset) -> Self {
		match asset {
			cf_chains::assets::btc::Asset::Btc =>
				Self { chain: ForeignChain::Bitcoin, asset: asset.into() },
		}
	}
}

#[derive(Serialize)]
#[serde(tag = "deposit_chain", rename_all = "snake_case")]
enum TransactionId {
	Bitcoin { hash: String },
	Ethereum { signature: SchnorrVerificationComponents },
	Polkadot { signature: String },
}

#[derive(Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
enum WitnessInformation {
	Deposit {
		deposit_chain_block_height: <AnyChain as Chain>::ChainBlockNumber,
		deposit_address: String,
		amount: NumberOrHex,
		asset: WitnessAsset,
	},
	Broadcast {
		broadcast_id: BroadcastId,
		tx_out_id: TransactionId,
	},
}

type DepositInfo<T> = (DepositWitness<T>, <T as Chain>::ChainBlockNumber);

impl From<DepositInfo<Ethereum>> for WitnessInformation {
	fn from((value, height): DepositInfo<Ethereum>) -> Self {
		Self::Deposit {
			deposit_chain_block_height: height,
			deposit_address: value.deposit_address.to_string(),
			amount: value.amount.into(),
			asset: value.asset.into(),
		}
	}
}

type BitcoinDepositInfo =
	(DepositWitness<Bitcoin>, <Bitcoin as Chain>::ChainBlockNumber, BitcoinNetwork);

impl From<BitcoinDepositInfo> for WitnessInformation {
	fn from((value, height, network): BitcoinDepositInfo) -> Self {
		Self::Deposit {
			deposit_chain_block_height: height,
			deposit_address: value.deposit_address.to_address(&network),
			amount: value.amount.into(),
			asset: value.asset.into(),
		}
	}
}

impl From<DepositInfo<Polkadot>> for WitnessInformation {
	fn from((value, height): DepositInfo<Polkadot>) -> Self {
		Self::Deposit {
			deposit_chain_block_height: height as u64,
			deposit_address: format!("0x{}", hex::encode(value.deposit_address.aliased_ref())),
			amount: value.amount.into(),
			asset: value.asset.into(),
		}
	}
}

async fn get_broadcast_id<I>(
	state_chain_client: &StateChainClient<()>,
	tx_out_id: &TransactionOutIdFor<state_chain_runtime::Runtime, I::Instance>,
) -> Option<BroadcastId>
where
	state_chain_runtime::Runtime: pallet_cf_broadcast::Config<I::Instance>,
	I: PalletInstanceAlias + 'static,
{
	let id = state_chain_client
		.storage_map_entry::<pallet_cf_broadcast::TransactionOutIdToBroadcastId<
			state_chain_runtime::Runtime,
			I::Instance,
		>>(state_chain_client.latest_unfinalized_block().hash, tx_out_id)
		.await
		.expect(STATE_CHAIN_CONNECTION)
		.map(|(broadcast_id, _)| broadcast_id);

	if id.is_none() {
		log::warn!("Broadcast ID not found for {:?}", tx_out_id);
	}

	id
}

#[derive(Clone, Deserialize, Debug)]
pub struct DepositTrackerSettings {
	eth: WsHttpEndpoints,
	dot: WsHttpEndpoints,
	state_chain_ws_endpoint: String,
	btc: HttpBasicAuthEndpoint,
}

#[derive(Parser, Debug, Clone, Default)]
#[clap(version = env!("SUBSTRATE_CLI_IMPL_VERSION"), version_short = 'v')]
pub struct TrackerOptions {
	#[clap(long = "eth.rpc.ws_endpoint")]
	eth_ws_endpoint: Option<String>,
	#[clap(long = "eth.rpc.http_endpoint")]
	eth_http_endpoint: Option<String>,
	#[clap(long = "dot.rpc.ws_endpoint")]
	dot_ws_endpoint: Option<String>,
	#[clap(long = "dot.rpc.http_endpoint")]
	dot_http_endpoint: Option<String>,
	#[clap(long = "state_chain.ws_endpoint")]
	state_chain_ws_endpoint: Option<String>,
	#[clap(long = "btc.rpc.http_endpoint")]
	btc_endpoint: Option<String>,
	#[clap(long = "btc.rpc.basic_auth_user")]
	btc_username: Option<String>,
	#[clap(long = "btc.rpc.basic_auth_password")]
	btc_password: Option<String>,
}

impl CfSettings for DepositTrackerSettings {
	type CommandLineOptions = TrackerOptions;

	fn load_settings_from_all_sources(
		config_root: String,
		_settings_dir: &str,
		opts: Self::CommandLineOptions,
	) -> Result<Self, ConfigError> {
		Self::set_defaults(Config::builder(), &config_root)?
			.add_source(Environment::default().separator("__"))
			.add_source(opts)
			.build()?
			.try_deserialize()
	}

	fn set_defaults(
		config_builder: ConfigBuilder<config::builder::DefaultState>,
		_config_root: &str,
	) -> Result<ConfigBuilder<config::builder::DefaultState>, ConfigError> {
		// These defaults are for a localnet setup
		config_builder
			.set_default("eth.ws_endpoint", "ws://localhost:8546")?
			.set_default("eth.http_endpoint", "http://localhost:8545")?
			.set_default("dot.ws_endpoint", "ws://localhost:9947")?
			.set_default("dot.http_endpoint", "http://localhost:9947")?
			.set_default("state_chain_ws_endpoint", "ws://localhost:9944")?
			.set_default("btc.http_endpoint", "http://127.0.0.1:8332")?
			.set_default("btc.basic_auth_user", "flip")?
			.set_default("btc.basic_auth_password", "flip")
	}

	fn validate_settings(
		&mut self,
		_config_root: &std::path::Path,
	) -> anyhow::Result<(), ConfigError> {
		Ok(())
	}
}

impl Source for TrackerOptions {
	fn clone_into_box(&self) -> Box<dyn Source + Send + Sync> {
		Box::new((*self).clone())
	}

	fn collect(&self) -> std::result::Result<Map<String, Value>, ConfigError> {
		let mut map: HashMap<String, Value> = HashMap::new();

		insert_command_line_option(&mut map, "eth.ws_endpoint", &self.eth_ws_endpoint);
		insert_command_line_option(&mut map, "eth.http_endpoint", &self.eth_http_endpoint);
		insert_command_line_option(&mut map, "dot.ws_endpoint", &self.dot_ws_endpoint);
		insert_command_line_option(&mut map, "dot.http_endpoint", &self.dot_http_endpoint);
		insert_command_line_option(
			&mut map,
			"state_chain_ws_endpoint",
			&self.state_chain_ws_endpoint,
		);
		insert_command_line_option(&mut map, "btc.http_endpoint", &self.btc_endpoint);
		insert_command_line_option(&mut map, "btc.basic_auth_user", &self.btc_username);
		insert_command_line_option(&mut map, "btc.basic_auth_password", &self.btc_password);

		Ok(map)
	}
}

async fn start(
	scope: &task_scope::Scope<'_, anyhow::Error>,
	settings: DepositTrackerSettings,
) -> anyhow::Result<()> {
	tracing_subscriber::FmtSubscriber::builder()
		.with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
		.try_init()
		.expect("setting default subscriber failed");
	let mut module = RpcModule::new(());

	let btc_tracker = witnessing::btc_mempool::start(scope, settings.btc.clone()).await;

	module.register_async_method("status", move |arguments, _context| {
		let btc_tracker = btc_tracker.clone();
		async move {
			arguments.parse::<Vec<String>>().map_err(Error::Call).and_then(|addresses| {
				btc_tracker
					.lookup_transactions(&addresses)
					.map_err(|err| jsonrpsee::core::Error::Custom(err.to_string()))
			})
		}
	})?;

	// Broadcast channel will drop old messages when the buffer is full to
	// avoid "memory leaks" due to slow receivers.
	const EVENT_BUFFER_SIZE: usize = 1024;
	let (witness_sender, _) =
		tokio::sync::broadcast::channel::<state_chain_runtime::RuntimeCall>(EVENT_BUFFER_SIZE);

	let (state_chain_client, env_params) =
		witnessing::start(scope, settings, witness_sender.clone()).await?;
	let btc_network = env_params.btc_network;

	module.register_subscription(
		"subscribe_witnessing",
		"s_witnessing",
		"unsubscribe_witnessing",
		move |_params, mut sink, _context| {
			let mut witness_receiver = witness_sender.subscribe();
			let state_chain_client = state_chain_client.clone();

			tokio::spawn(async move {
				while let Ok(event) = witness_receiver.recv().await {
					use pallet_cf_broadcast::Call as BroadcastCall;
					use pallet_cf_ingress_egress::Call as IngressEgressCall;
					use state_chain_runtime::RuntimeCall::*;

					// rustfmt chokes when formatting this macro.
					// See: https://github.com/rust-lang/rustfmt/issues/5404
					#[rustfmt::skip]
					macro_rules! send {
						($value:expr) => {
							if let Ok(false) = sink.send($value) {
								log::debug!("Subscription is closed");
								return
							}
						};
					}

					match event {
						EthereumIngressEgress(IngressEgressCall::process_deposits {
							deposit_witnesses,
							block_height,
						}) =>
							for witness in deposit_witnesses as Vec<DepositWitness<Ethereum>> {
								let info = WitnessInformation::from((witness, block_height));
								send!(&info);
							},
						BitcoinIngressEgress(IngressEgressCall::process_deposits {
							deposit_witnesses,
							block_height,
						}) =>
							for witness in deposit_witnesses as Vec<DepositWitness<Bitcoin>> {
								let info =
									WitnessInformation::from((witness, block_height, btc_network));
								send!(&info);
							},
						PolkadotIngressEgress(IngressEgressCall::process_deposits {
							deposit_witnesses,
							block_height,
						}) =>
							for witness in deposit_witnesses as Vec<DepositWitness<Polkadot>> {
								let info = WitnessInformation::from((witness, block_height));
								send!(&info);
							},
						EthereumBroadcaster(BroadcastCall::transaction_succeeded {
							tx_out_id,
							..
						}) => {
							let broadcast_id =
								get_broadcast_id::<Ethereum>(&state_chain_client, &tx_out_id).await;

							if let Some(broadcast_id) = broadcast_id {
								send!(&WitnessInformation::Broadcast {
									broadcast_id,
									tx_out_id: TransactionId::Ethereum { signature: tx_out_id }
								})
							}
						},
						BitcoinBroadcaster(BroadcastCall::transaction_succeeded {
							tx_out_id,
							..
						}) => {
							let broadcast_id =
								get_broadcast_id::<Bitcoin>(&state_chain_client, &tx_out_id).await;

							if let Some(broadcast_id) = broadcast_id {
								send!(&WitnessInformation::Broadcast {
									broadcast_id,
									tx_out_id: TransactionId::Bitcoin {
										hash: format!("0x{}", hex::encode(tx_out_id))
									}
								})
							}
						},
						PolkadotBroadcaster(BroadcastCall::transaction_succeeded {
							tx_out_id,
							..
						}) => {
							let broadcast_id =
								get_broadcast_id::<Polkadot>(&state_chain_client, &tx_out_id).await;

							if let Some(broadcast_id) = broadcast_id {
								send!(&WitnessInformation::Broadcast {
									broadcast_id,
									tx_out_id: TransactionId::Polkadot {
										signature: format!(
											"0x{}",
											hex::encode(tx_out_id.aliased_ref())
										)
									}
								})
							}
						},

						EthereumIngressEgress(_) |
						BitcoinIngressEgress(_) |
						PolkadotIngressEgress(_) |
						System(_) |
						Timestamp(_) |
						Environment(_) |
						Flip(_) |
						Emissions(_) |
						Funding(_) |
						AccountRoles(_) |
						Witnesser(_) |
						Validator(_) |
						Session(_) |
						Grandpa(_) |
						Governance(_) |
						Reputation(_) |
						TokenholderGovernance(_) |
						EthereumChainTracking(_) |
						BitcoinChainTracking(_) |
						PolkadotChainTracking(_) |
						EthereumVault(_) |
						PolkadotVault(_) |
						BitcoinVault(_) |
						EthereumThresholdSigner(_) |
						PolkadotThresholdSigner(_) |
						BitcoinThresholdSigner(_) |
						EthereumBroadcaster(_) |
						PolkadotBroadcaster(_) |
						BitcoinBroadcaster(_) |
						Swapping(_) |
						LiquidityProvider(_) |
						LiquidityPools(_) => {},
					};
				}
			});
			Ok(())
		},
	)?;

	scope.spawn(async {
		let server = ServerBuilder::default().build("0.0.0.0:13337".parse::<SocketAddr>()?).await?;
		let addr = server.local_addr()?;
		log::info!("Listening on http://{}", addr);
		server.start(module)?.stopped().await;
		// If the server stops for some reason, we return
		// error to terminate other tasks and the process.
		Err(anyhow::anyhow!("RPC server stopped"))
	});

	Ok(())
}

#[tokio::main]
async fn main() -> anyhow::Result<()> {
	let settings = DepositTrackerSettings::load_settings_from_all_sources(
		// Not using the config root or settings dir.
		"".to_string(),
		"",
		TrackerOptions::parse(),
	)?;

	task_scope::task_scope(|scope| async move { start(scope, settings).await }.boxed()).await
}
