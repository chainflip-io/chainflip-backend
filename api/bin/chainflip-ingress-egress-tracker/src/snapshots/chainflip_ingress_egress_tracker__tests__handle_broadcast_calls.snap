---
source: api/bin/chainflip-ingress-egress-tracker/src/main.rs
expression: "store.storage.get(\"broadcast:ethereum:1\").unwrap()"
---
{"broadcast_id":1,"tx_out_id":{"deposit_chain":"Ethereum","signature":{"k_times_g_address":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"s":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]}},"type":"broadcast"}
